rfcs/react-components/convergence/portal-compat.md

```diff
@@ -0,0 +1,166 @@
+# RFC: Portal compatibility
+
+[@layershifter](https://github.com/layershifter)
+
+## Summary
+
+This RFC proposes introduce compatibility between components in different Fluent UI versions that use [`ReactDOM.createPortal()`](https://reactjs.org/docs/portals.html) via:
+
+- usage of same values for `z-index`
+- usage of separate React context to assign CSS class that handles CSS variables from Fluent UI React v9
+
+## Problem statement
+
+Once components from Fluent UI React v9 (`@fluentui/react-components`) are put inside popovers & dialogs from Fluent UI React v8 (`@fluentui/react`) or Fluent UI React v0 (`@fluentui/react-northstar`) they don't have proper styles 💥
+
+Examples are below:
+
+- v8 & v9: https://codesandbox.io/s/v8-v9-no-compat-2nnt7
+- v0 & v9: https://codesandbox.io/s/v0-v9-no-compat-vw9e2
+
+Another problem is related to `z-index` values. Fluent UI React v8 (`@fluentui/react`) defines values that win over everything:
+
+```tsx
+export namespace ZIndexes {
+  export const Nav: number = 1;
+  /**
+   * @deprecated Do not use
+   */
+  export const ScrollablePane: number = 1;
+  export const FocusStyle: number = 1;
+  export const Coachmark: number = 1000;
+  export const Layer: number = 1000000;
+  export const KeytipLayer: number = 1000001;
+}
+```
+
+> Check [style-utilities/src/styles/zIndexes.ts](https://github.com/microsoft/fluentui/blob/36df8cdbdc8dd4b5989685f89b7d339ee5774bf7/packages/style-utilities/src/styles/zIndexes.ts#L1-L11)
+
+As v9 does not use such values components that use portals (`Popover`, `Menu`, `Tooltip`) appear **under** components from v8.
+
+Demo: https://codesandbox.io/s/v8layer-v9popover-q33b37
+
+## Detailed Design or Proposal
+
+### Use same `z-index` values
+
+- We can't change values in Fluent UI React v8 as this will be a breaking change
+- We want to make our components visible
+
+The simplest option that satisfies these requirements is to use same values in Fluent UI React v9.
+
+#### Pros and Cons
+
+- 👎 `z-index=1000000` is ugly
+- 👍 it's predictable and works without any magic
+- 👍 no breaking changes
+
+### Introduce compat React Context
+
+Introduce Pub/Sub context where:
+
+- Publisher adds/modifies classes on passed elements
+- Subscriber registers portal elements and does not know anything about implementation
+
+In the prototype the publisher component is called `CompatHostV9`:
+
+- it reads current classes from `useThemeClassName()`
```

Comment:
apps could also just create their own class name instead of `useThemeClassName` (i.e. use `useThemeStyleTag` instead.

Maybe we should include that as an option and also list pros/cons

Reply:
I am not sure that it's a realistic case... `useThemeClassName()` and matching context are still required to be used to make Fluent components working otherwise heavy recomposition of every component that uses `Portal` will be required.

For custom components I still see sense to use our `Portal` instead of custom ones (it handles SSR). IMO if customers are creating custom portals and do not use `useThemeClassName()` to get classes with CSS variables they have bigger problem than compatibility 🐱